mlflow/models/evaluation/default_evaluator.py

```diff
@@ -1273,71 +1315,18 @@ def _generate_model_predictions(self):
-            os.environ["TIKTOKEN_CACHE_DIR"] = ""
-
-            encoding = tiktoken.get_encoding("cl100k_base")
-
-            def compute_num_tokens(y_pred):
-                # parse out the output from y_pred
-                if isinstance(y_pred, pd.DataFrame):
-                    output = y_pred.iloc[0, 0]
-                elif isinstance(y_pred, (np.ndarray, list)):
-                    output = y_pred[0]
-                elif isinstance(y_pred, pd.Series):
-                    output = y_pred.iloc[0]
-                # if output is string-like, tokenize it and get the number of tokens
-                if isinstance(output, str):
-                    return len(encoding.encode(output))
-                else:
-                    return None
-
-            if len(X_copy) == 0:
-                raise ValueError("Empty input data")
-
-            is_dataframe = isinstance(X_copy, pd.DataFrame)
-
-            for row in X_copy.iterrows() if is_dataframe else enumerate(X_copy):
-                i, row_data = row
-                single_input = row_data.to_frame().T if is_dataframe else row_data
-                start_time = time.time()
-                y_pred = self.model.predict(single_input)
-                end_time = time.time()
-                num_tokens_list.append(compute_num_tokens(y_pred))
-                pred_latencies.append(end_time - start_time)
-                y_pred_list.append(y_pred)
-
-            # Aggregate all predictions into model_predictions
-            sample_pred = y_pred_list[0]
-            if isinstance(sample_pred, pd.DataFrame):
-                model_predictions = pd.concat(y_pred_list)
-            elif isinstance(sample_pred, np.ndarray):
-                model_predictions = np.concatenate(y_pred_list, axis=0)
-            elif isinstance(sample_pred, list):
-                model_predictions = sum(y_pred_list, [])
-            # handle if sample_pred is a pandas series
-            elif isinstance(sample_pred, pd.Series):
-                model_predictions = pd.concat(y_pred_list)
+            if compute_latency:
+                model_predictions = predict_with_latency(X_copy)
             else:
-                raise MlflowException(
-                    message=f"Unsupported prediction type {type(sample_pred)} for model type "
-                    f"{self.model_type}.",
-                    error_code=INVALID_PARAMETER_VALUE,
-                )
+                model_predictions = self.model.predict(X_copy)
 
-            self.metrics_values.update({_LATENCY_METRIC_NAME: MetricValue(scores=pred_latencies)})
-            self.metrics_values.update(
-                {_TOKEN_COUNT_METRIC_NAME: MetricValue(scores=num_tokens_list)}
-            )
         else:
-            model_predictions = self.model.predict(self.X.copy_to_avoid_mutation())
+            X_copy = self.X.copy_to_avoid_mutation()
+            if compute_latency:
+                model_predictions = predict_with_latency(X_copy)
```

Comment:
Can we de dedup this to be at the top level that if we have

if compute_latency:
                model_predictions = predict_with_latency(X_copy)
 else if regression
 else if qa, text, etc
 
 Does that make sense?